### PR TITLE
Make ShufflerDataPipe deterministic for SP & MP DataLoader

### PR DESCRIPTION
**Pull request by ejguan**
This is the first PR to make DataPipe deterministic. See: https://github.com/pytorch/data/issues/302

Users should be able to use `torch.manual_seed(seed)` to control the shuffle order for the following cases:
- Directly over `DataPipe`
- For single-process DataLoader
- Multiprocessing DataLoader

Unfortunately, for distributed training, users have to run `apply_shuffle_seed` manually to make sure all distributed processes having the same order of shuffle.

### CODE REVIEW

File: test/test_datapipe.py [hunk 1]
```diff
@@ -1469,26 +1472,53 @@ def test_stream_reader_iterdatapipe(self):
...
+                mp_ctx = "spawn" if num_workers > 0 else None
+                dl = DataLoader(
+                    shuffle_dp,
+                    num_workers=num_workers,
+                    shuffle=True,
+                    multiprocessing_context=mp_ctx,
+                    worker_init_fn=_worker_init_fn
+                )
+                for epoch in range(2):
+                    torch.manual_seed(123)
```
Comment (VitalyFedyunin) [inline test/test_datapipe.py:1501]: Worth testing that lists are different w/o manual seed.

**Review by VitalyFedyunin — APPROVED** [review]

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 2]
```diff
@@ -94,65 +98,82 @@ def __init__(self,
...
     def reset(self) -> None:
-        self.buffer = []
+        if not self._restored:
+            self._buffer = []
+            if self._enabled and self._seed is None:
+                self._seed = int(torch.empty((), dtype=torch.int64).random_().item())
+            self._rng.seed(self._seed)
+            self._seed = None
+        else:
+            self._restored = False
```
[the remark this answers is not in the document]
Comment (NivekT) [reply torch/utils/data/datapipes/iter/combinatorics.py:151]: That's a good call. I think we should do that across all DataPipes that have a custom `reset()` method.

File: torch/utils/data/datapipes/iter/combinatorics.py [hunk 3]
```diff
@@ -94,65 +98,82 @@ def __init__(self,
                  unbatch_level: int = 0
```
Comment (NivekT) [inline torch/utils/data/datapipes/iter/combinatorics.py:98]: Should we accept `seed` as an optional argument here? What's downside if we do that?
It looks like the intention here for the users to pass RNG from DataLoader rather than directly setting one?

**Review by NivekT — APPROVED** [review]
LGTM! We will have to update the docs/examples at some point but not necessary in this PR.

File: torch/utils/data/datapipes/iter/combinatorics.py (hunk 3 above)
Comment (ejguan) [reply torch/utils/data/datapipes/iter/combinatorics.py:98]: Since users want to run shuffle with different order per epoch, providing a single seed is not enough.
You can treat the RNG in DataLoader as the single source of SeedGenerator. All DataPipe with random op should fetch seeds from it.
It's easier to maintain.

**Comment by NicolasHug** [comment]
> Unfortunately, for distributed training, users have to run apply_shuffle_seed manually to make sure all distributed processes having the same order of shuffle.

@ejguan could you please provide an example of what users will have to write for the shuffling to be the same across distributed workers? It seems that the `apply_shuffle_seed` takes a Generator as input, but it's not obvious to me yet how to make sure that this generator is the same across workers while still supporting:

- determinism
- maximal entroypy, e.g. different shuffling for different program execution.

**Comment by ejguan** [comment]
@NicolasHug 

> It seems that the `apply_shuffle_seed` takes a Generator as input, but it's not obvious to me yet how to make sure that this generator is the same across workers while still supporting:

The generator used by `apply_shuffle_seed` in each worker process is created from the same `base_seed`. And, the `base_seed` is generated in main process and using `torch.randint`, which can be determined by `torch.manual_seed()`.

For distributed workers, users need to explicitly specify the same seed either using `torch.manual_seed` in the training script, or using `apply_shuffle_seed` in the `worker_init_fn` with the RNGs with the same seed.

**Comment by NicolasHug** [comment]
Thanks for the details Erjia. Could we possibly have this synchronization of the seed across DDP workers done automatically by default?
These are low-level implementation details that, ideally, users shouldn't have to care about IMHO.
